dolfin_adjoint_common/blocks/assembly.py

```diff
@@ -18,6 +18,41 @@ def __init__(self, form, ad_block_tag=None):
     def __str__(self):
         return f"assemble({ufl2unicode(self.form)})"
 
+    def compute_action_adjoint(self, adj_input, arity_form, form=None, c_rep=None, space=None, dform=None):
+        if arity_form == 0:
+            if dform is None:
+                dc = self.backend.TestFunction(space)
+                dform = self.backend.derivative(form, c_rep, dc)
+            dform_vector = self.compat.assemble_adjoint_value(dform)
+            return adj_input * dform_vector, dform
+        elif arity_form == 1:
+            if dform is None:
+                dc = self.backend.TrialFunction(space)
+                dform = self.backend.derivative(form, c_rep, dc)
+            # Get the Function
+            adj_input = adj_input.function
+            # Symbolic operators such as action/adjoint require derivatives to have been expanded beforehand.
+            # However, UFL doesn't support expanding coordinate derivatives of Coefficients in physical space,
+            # implying that we can't symbolically take the action/adjoint of the Jacobian for SpatialCoordinates.
+            # -> Workaround: Apply action/adjoint numerically (using PETSc).
+            if not isinstance(c_rep, self.backend.SpatialCoordinate):
+                # Symbolically compute: (dform/dc_rep)^* * adj_input
+                adj_output = self.backend.action(self.backend.adjoint(dform), adj_input)
+                adj_output = self.compat.assemble_adjoint_value(adj_output)
+            else:
+                # Get PETSc matrix
+                dform_mat = self.compat.assemble_adjoint_value(dform).petscmat
+                # Adjoint (hermitian transpose)
+                dform_mat.hermitianTranspose()
+                # Action
+                adj_output = self.backend.Function(space)
+                with adj_input.dat.vec_ro as v_vec:
+                    with adj_output.dat.vec as res_vec:
+                        dform_mat.mult(v_vec, res_vec)
```

Comment:
Delete line 46 and use `.multHermitian` here instead. This avoids having to actually make an adjoint matrix.

Reply:
Thanks!